lib/service/awsoidc.go

```diff
@@ -0,0 +1,263 @@
+}
+
+func (process *TeleportProcess) updateDeployServiceAgents(ctx context.Context, authClient *auth.Client) error {
+	cmc, err := authClient.GetClusterMaintenanceConfig(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var criticalEndpoint string
+	if automaticupgrades.GetChannel() != "" {
+		criticalEndpoint, err = url.JoinPath(automaticupgrades.GetChannel(), "critical")
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	critical, err := automaticupgrades.Critical(process.GracefulExitContext(), criticalEndpoint)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if !withinUpgradeWindow(cmc, process.Clock) && !critical {
+		return nil
+	}
+
+	teleportVersion, err := getStableTeleportVersion(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	issuer, err := awsoidc.IssuerFromPublicAddress(process.proxyPublicAddr().Addr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	clusterNameConfig, err := authClient.GetClusterName()
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	clusterName := clusterNameConfig.GetClusterName()
+
+	databases, err := authClient.GetDatabases(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	awsRegions := make(map[string]interface{})
+	for _, database := range databases {
+		if database.IsAWSHosted() && database.IsRDS() {
+			awsRegions[database.GetAWS().Region] = nil
+		}
+	}
+
+	var resources []types.Integration
+	var nextKey string
+	for {
+		igs, nextKey, err := authClient.ListIntegrations(ctx, 0, nextKey)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		resources = append(resources, igs...)
+		if nextKey == "" {
+			break
+		}
+	}
+
+	limit := rate.NewLimiter(rate.Every(updateDeployAgentsRateLimit), 1)
+	for _, ig := range resources {
+		spec := ig.GetAWSOIDCIntegrationSpec()
+		if spec == nil {
+			continue
+		}
+		integrationName := ig.GetName()
+
+		for region := range awsRegions {
+			if err := limit.Wait(ctx); err != nil {
+				return trace.Wrap(err)
+			}
+
+			token, err := authClient.GenerateAWSOIDCToken(ctx, types.GenerateAWSOIDCTokenRequest{
+				Issuer: issuer,
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			req := &awsoidc.AWSClientRequest{
+				IntegrationName: ig.GetName(),
+				Token:           token,
+				RoleARN:         spec.RoleARN,
+				Region:          region,
+			}
+
+			deployServiceClient, err := awsoidc.NewDeployServiceClient(ctx, req, authClient)
+			if err != nil {
+				process.log.Warningf("Failed to update deploy service agents: %v", err)
+				continue
+			}
+
+			ownershipTags := map[string]string{
+				types.ClusterLabel:     clusterName,
+				types.OriginLabel:      types.OriginIntegrationAWSOIDC,
+				types.IntegrationLabel: integrationName,
+			}
+
+			// Acquire a lease for the region + integration before attempting to update the deploy service agent.
+			// If the lease cannot be acquired, the update is already being handled by another instance.
+			semLock, err := authClient.AcquireSemaphore(ctx, types.AcquireSemaphoreRequest{
```

Comment:
It looks like the semaphore is only released explicitly in case of error below. Should we release it in case of success as well?

Reply:
If we release the semaphore on success, other instances can acquire the lock in the same update interval. This means that the update may be attempted for the same Deploy Service Agents by multiple instances. This isn't necessarily a problem because the update will return early after comparing the current Teleport version and the stable version, but it does make a few API calls before then. So holding onto the semaphore until the next update interval will help us reduce the number of API calls made.